src/core/lwip/tcp.c

```diff
@@ -148,6 +148,14 @@ static err_t tcp_close_shutdown(struct tcp_pcb *pcb, u8_t rst_on_unacked_data)
 {
     err_t err;
 
+    if (get_tcp_state(pcb) == SYN_RCVD) {
+        // according to the RFC, in case we get a SYN and no more data
+        // we should just close w/o FIN or RST
+        tcp_pcb_purge(pcb);
+        set_tcp_state(pcb, CLOSED);
+        return ERR_OK;
+    }
+
```

Comment:
The RFC states that FIN needs to be sent in this case. So, looks like original code makes sense.

[section 3.10.4](https://datatracker.ietf.org/doc/html/rfc9293#name-close-call)
```
3.10.4.  CLOSE Call
...
SYN-RECEIVED STATE

If no SENDs have been issued and there is no pending data to send, then form a FIN segment and send it, and enter FIN-WAIT-1 state; otherwise, queue for processing after entering ESTABLISHED state.
```

Reply:
You're right @pasis.
If we get a timeout we should just abandon.
If an app closes an accepted socket, we should initiate FIN flow.

2 issues:
1. Currently, on vNext - closing an accepted socket just sends a fin and closes the socket (calls internal thread with UNREGISTER_TCP_SOCKET_TIMER_AND_DELETE which destructs it). this doesn't conform too to the RFC!
2. there's no real distinguishing between app close to xlio initiating the close. See `handle_incoming_handshake_failure` - it actively calls the close syscall if we get a timeout - hence, we can't distinguish really between the two cases.

This makes the bug a much harder to fix - but I'm up to the challenge :) let me know if you got me and we're on the same page.

Reply:
@pasis 
Actually, I instrumented the code, adding this to tcp.c line 760:
printf("DEBUG: tcp_slowtmr PATH1 BEFORE purge: state=%d, unsent=%p unacked=%p\n",
                   get_tcp_state(pcb), pcb->unsent, pcb->unacked);

it validated we are aborting due to being "too long in SYN-RCVD" - we are not in CLOSE flow (3.10.4).

I've tried to look online what to do on timeout on SYN-RCVD, but it's not well-defined.
the linux kernel though simply abandons w/o sending RST, w/o sending FIN, as we do here - so I believe this solution is correct

Reply:
but this instrumentation does show the fix is not quite right - let me alter it